libraries/Microsoft.Bot.Builder/BotServer.cs

```diff
@@ -0,0 +1,143 @@
+{
+    public abstract class BotServer
+    {
+        protected readonly Middleware.MiddlewareSet _middlewareSet = new Middleware.MiddlewareSet();
+
+        public BotServer() : base()
+        {
+            this.RegisterMiddleware(new Middleware.BindOutoingResponsesMiddlware());
+            this.RegisterMiddleware(new Middleware.TemplateManager());
+        }
+
+        /// <summary>
+        /// Register middleware with the bot
+        /// </summary>
+        /// <param name="middleware"></param>
+        public void RegisterMiddleware(IMiddleware middleware)
+        {
+            _middlewareSet.Use(middleware);
+        }
+
+        /// <summary>
+        /// implement send an activity to the conversation
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name=""></param>
+        /// <returns></returns>
+        protected abstract Task SendActivityImplementation(IBotContext context, IActivity activity);
+
+        /// <summary>
+        /// Implement updating an activity in the conversation
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="activity"></param>
+        /// <returns></returns>
+        protected abstract Task<ResourceResponse> UpdateActivityImplementation(IBotContext context, IActivity activity);
+
+        /// <summary>
+        /// Implement deleting an activity in the conversation
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="conversationId"></param>
+        /// <param name="activityId"></param>
+        /// <returns></returns>
+        protected abstract Task DeleteActivityImplementation(IBotContext context, string conversationId, string activityId);
+
+
+        /// <summary>
+        /// Implement createconversation semantics
+        /// </summary>
+        /// <returns></returns>
+        protected abstract Task CreateConversationImplementation();
+
+
+        protected async Task RunPipeline(IBotContext context, Func<IBotContext, Task> callback = null)
+        {
+            BotAssert.ContextNotNull(context);
+
+            System.Diagnostics.Trace.TraceInformation($"Middleware: Beginning Pipeline for {context.ConversationReference.ActivityId}");
+
+            // Call any registered Middleware Components looking for ContextCreated()
+            await _middlewareSet.ContextCreated(context).ConfigureAwait(false);
+
+            // Call any registered Middleware Components looking for ReceiveActivity()
+            if (context.Request != null)
+            {
+                await _middlewareSet.ReceiveActivity(context).ConfigureAwait(false);
+            }
+
+            // call back to caller
+            if (callback != null)
+            {
+                await callback(context).ConfigureAwait(false);
+            }
+
+            // Call any registered Middleware Components looking for SendActivity()
+            if (context.Responses != null && context.Responses.Any())
```

Comment:
Refer to #144

Given the current factoring (of state in middleware) this would still be an issue because of this line